docs/gguf.md

```diff
@@ -0,0 +1,576 @@
+    // The value is a 64-bit signed little-endian integer.
+    GGUF_METADATA_VALUE_TYPE_INT64 = 11,
+    // The value is a 64-bit IEEE754 floating point number.
+    GGUF_METADATA_VALUE_TYPE_FLOAT64 = 12,
+}
+
+// A string in GGUF.
+struct gguf_string_t {
+    // The length of the string, in bytes.
+    uint64_t len;
+    // The string as a UTF-8 non-null-terminated string.
+    char string[len];
+}
+
+union gguf_metadata_value_t {
+    uint8_t uint8;
+    int8_t int8;
+    uint16_t uint16;
+    int16_t int16;
+    uint32_t uint32;
+    int32_t int32;
+    float float32;
+    uint64_t uint64;
+    int64_t int64;
+    double float64;
+    bool bool_;
+    gguf_string_t string;
+    struct {
+        // Any value type is valid, including arrays.
+        gguf_metadata_value_type type;
+        // Number of elements, not bytes
+        uint64_t len;
+        // The array of values.
+        gguf_metadata_value_t array[len];
+    } array;
+};
+
+struct gguf_metadata_kv_t {
+    // The key of the metadata. It is a standard GGUF string, with the following caveats:
+    // - It must be a valid ASCII string.
+    // - It must be a hierarchical key, where each segment is `lower_snake_case` and separated by a `.`.
+    // - It must be at most 2^16-1/65535 bytes long.
+    // Any keys that do not follow these rules are invalid.
+    gguf_string_t key;
+
+    // The type of the value.
+    // Must be one of the `gguf_metadata_value_type` values.
+    gguf_metadata_value_type value_type;
+    // The value.
+    gguf_metadata_value_t value;
+};
+
+struct gguf_header_t {
+    // Magic number to announce that this is a GGUF file.
+    // Must be `GGUF` at the byte level: `0x47` `0x47` `0x55` `0x46`.
+    // Your executor might do little-endian byte order, so it might be
+    // check for 0x46554747 and letting the endianness cancel out.
+    // Consider being *very* explicit about the byte order here.
+    uint32_t magic;
+    // The version of the format implemented.
+    // Must be `2` for version described in this spec.
+    //
+    // This version should only be increased for structural changes to the format.
+    // Changes that do not affect the structure of the file should instead update the metadata
+    // to signify the change.
+    uint32_t version;
+    // The number of tensors in the file.
+    // This is explicit, instead of being included in the metadata, to ensure it is always present
+    // for loading the tensors.
+    uint64_t tensor_count;
+    // The number of metadata key-value pairs.
+    uint64_t metadata_kv_count;
+    // The metadata key-value pairs.
+    gguf_metadata_kv_t metadata_kv[metadata_kv_count];
+};
+
+struct gguf_tensor_info_t {
+    // The name of the tensor. It is a standard GGUF string, with the caveat that
+    // it must be at most 64 bytes long.
+    gguf_string_t name;
+    // The number of dimensions in the tensor.
+    // Currently at most 4, but this may change in the future.
+    uint32_t n_dimensions;
```

Comment:
This should also be uint64_t

Reply:
I thought about this, but I think we'd have much larger problems if there are tensors with more than 4 billion dimensions. I can do it if you really want, but I don't know if we should suggest that this is even a possibility, given the current maximum for this value is 4.

Reply:
No, this is fine as `uint32_t`

Reply:
I guess uint32_t will have to do for n_dims.